samples/userspace/shared_mem/src/main.c

```diff
@@ -192,6 +192,7 @@ void main(void)
 
 	k_thread_start(&ct_thread);
 	k_sem_give(&allforone);
+	k_yield();
```

Comment:
> As it happened, the implementation of k_sem_take() always acted like a
scheduling point, so the test never got stuck.  But this behavior was
never documented, 

From the docs:

	There are various points in time when the scheduler is given an
	opportunity to change the identity of the current thread.  These points
	are called **reschedule points**. Some potential reschedule points are:
	
	- transition of a thread from running state to a suspended or waiting
	  state, for example by :c:func:`k_sem_take` or :c:func:`k_sleep`.

>  and newer synchronization mechanisms like zync don't
implement it like that (in fact zync is a 100% kernel free atomic
operation that can't possibly block in the uncontended case).

Isn't this breaking existing and "documented" behavior? 
additionally, I think it is the wrong approach to be changing tests to make them work with a new implementation that might change bahviour that is documented. This might have negative impact on user who expect this behaviour. The tests is all we have to verify that we are not breaking or changing existing behavior. 

IMO changes here (beside the obvious bug fixes) need to put on hold until the new implementation has been reviewed, discussed and there is agreement we want to go this path and change bahviour in areas in a backward complatible way.

Reply:
Those docs are correct, btw.  Zync will reschedule just fine when there are scheduler state changes, just like the docs say.  The issue is that semaphores were unconditional scheduling points before, and aren't now (and can't be, and we want that).